x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/LifecycleAction.java

```diff
@@ -29,6 +30,15 @@
      */
     List<Step> toSteps(Client client, String phase, @Nullable Step.StepKey nextStepKey);
 
+    /**
+     * 
+     * @param phase
+     *            the name of the phase this action is being executed within
+     * @return the {@link StepKey}s for the steps which will be executed in this
+     *         action
+     */
+    List<StepKey> toStepKeys(String phase);
```

Comment:
should we find a way to do this so that either `toSteps` is derived from `toStepKeys`, or vice-versa? The separation of these in ways that need to be kept consistent sounds like it may be difficult to keep in sync

Reply:
Yeah I thought about this at the time but the problem is that in order to derive the steps from the step keys we would end up having some ugly if-else logic to iterate through the keys returned by `toStepKeys` and for each try to work out which step it is based on and build that step. The reverse of building the stepKeys based on the output of `toSteps` meant passing in a `null` client in production code which I also didn't like because it could potentially cause NPE's. In order to try to keep consistency between the two methods `AbstractActionTestCase` has a method which create a random instance of the action under test and makes sure that the step keys from the steps generated by `toSteps()` are the same as the step keys created by `toStepKeys()`: https://github.com/elastic/elasticsearch/pull/32283/files#diff-99c08c350d9cf07e3a5366ab534f80f7R28

Reply:
OK, then I guess that is the best we can do for now. I do not see a better way at the moment either